cxx4opencl/address_spaces.txt

```diff
@@ -1,135 +1,207 @@
+__private int * ptr = addrspace_cast<__private int*>(privptr); // legal, no op.
+float * privptrfloat = reinterpret_cast<float*>(ptr); // legal.
+
+// disjoint address space conversion.
+__constant int * constptr = addrspace_cast<__constant int*>(privptr); // illegal.
+------------
 
 [[references]]
 ==== References
 
 Reference types can be qualified with an address space.
 
 [source,cpp]
-----------
+------------
 __private int & ref = ...; // references int in __private address space.
-----------
+------------
 
-By default references refer to generic address space objects, except
+By default references refer to generic address space objects if generic
+address space is supported or private address space otherwise, except
 for dependent types that are not template specializations (see
-<<addrspace-deduction, _Deduction_>>). Address space compatibility checks
-are performed when references are bound to values. The logic follows the
-rules from address space pointer conversion (`OpenCL v2.0 s6.5.5`).
+<<addrspace-deduction, _Deduction_>>).
 
+[source,cpp]
+------------
+int & ref = ...; // references int in generic address space if it is
+                 // supported otherwise in __private address space.
+------------
+
+Address space compatibility checks are performed when references are
+bound to values. The logic follows the rules from address space pointer
+conversion (`OpenCL C 3.0 s6.7.9`).
+
+[source,cpp]
+------------
+void f(float &ref, __global float &globref) {
+  const int& tmp = ref; // legal - reference to generic/__private address space object can
+                        // bind to a temporary object created in __private  address space.
+
+  __global const int& globtmp = globref; // error reference to global address space object cannot bind
+                                         // to a temporary object created in __private  address space.
+}
+------------
 
 [[addrspace-deduction]]
-==== Deduction & Default address space
+====  Address space inference
 
 This section details what happens if address spaces for types are not
 provided in the source code explicitly. Most of the logic for address space
-deduction (i.e. default address space) follows rules from `OpenCL v2.0 s6.5`.
+inference (i.e. default address space) follows rules from `OpenCL C 3.0 s6.7.8`.
 
 References inherit rules from pointers and therefore refer to generic
-address space objects by default (see <<references, _References_>>).
+address space objects by default (see <<references, _References_>>) if
+generic address space is supported, otherwise they refer to private
+address space objects.
 
-Class static data members are deduced to `+__global+` address space.
+Class static data members are deduced to `+__global+` address space. Note,
+that if implementation of {cpp} for OpenCL 2021 does not support program
+scope variables (i.e. `+__opencl_c_program_scope_global_variables+` feature
+from `OpenCL C 3.0 s6.1`) the address space qualifier must be  specified
```

Comment:
```suggestion
from `OpenCL C 3.0 s6.1` is disabled) the address space qualifier must be specified
```

Reply:
I would like to avoid using "disabled" as this can be confused with pragma enabling/disabling mechanism for extensions. How about:
```suggestion
from `OpenCL C 3.0 s6.1` is unsupported) the address space qualifier must be specified
```

Reply:
Unsupported is indeed better!